### DISCUSSION TITLE
ER: struct_keys()

### DISCUSSION THREAD

**Discussion by pkoppstein** [duckdb/duckdb#13454] [Ideas]
DuckDB has map_keys(map), and the JSON extension gives us json_keys().
These highlight what I understand to be a gap in DuckDB itself: the
equivalent for getting the "keys" of a STRUCT.

This gap is consequential, because it causes unwanted complexities and
inefficiencies in the programmatic handling of structs, which are
ubiquitous in DuckDB.  

For example, if t is a table with two columns,
we might see:
```
select t from t;
┌──────────────────────────────┐
│              t               │
│ struct(i integer, j integer) │
├──────────────────────────────┤
│ {'i': 0, 'j': 1}             │
│ {'i': 1, 'j': 2}             │
│ {'i': 2, 'j': 3}             │
└──────────────────────────────┘
```

At present, to make full programmatic use of this "horizontal" view, we have to convert the
struct to JSON.  Consider, for example, this MACRO for dynamically computing a list of the column
names of a table:

```
create or replace macro cols(t) as (select json_keys(j) from (select t::json j from t));
```

This is nice but horrendously inefficient because the entire table must be converted to JSON,
which would not be needed if we had struct_keys().

**Comment by Maxxen** [comment]
Im not sure how this is supposed to work...? Because the keys are part of the type definition of a struct column, `struct_keys()` as a scalar function will return the same result for each row, and not return anything at all for an empty table. How is that useful?

**Reply by soerenwolfers** [reply]
Doesn't the same argument work for `pg_typeof(x)`?

**Reply by Maxxen** [reply]
Exactly, and I dont think typeof is that useful either, at least I cant really see how you would use the result for further processing and/or type manipulation. If struct_keys is ment to fit into the same use case I would rather consider adding something like a "type_parts" that can "deconstruct" and give you the type "schema" for any nested type.

**Comment by soerenwolfers** [comment]
you could get the keys via
```sql
SELECT column_name FROM (DESCRIBE SELECT  my_struct.* FROM t)
```

**Comment by pkoppstein** [comment]
@soerenwolfers - 

1. What if there’s no table (t)?

2. Don’t forget the restriction:

Binder Error: subqueries in lambda expressions are not supported

**Reply by soerenwolfers** [reply]
If you give examples for both of these, I'll see if I can find workarounds

**Comment by pkoppstein** [comment]
@soerenwolfers - I fear that any examples I give will be misconstrued in some way.  In particular, I'm not  looking for workarounds.  If your point is that a more worthy addition to DuckDB would be struct_transform() (in analogy to list_transform()), then I might agree.
   
However, if an example will help rather than misdirect, here's one.

Suppose I want to write a generic function that will handle any table, and in particular tables in which some columns are VARCHARs. Suppose I want to extract all those strings into a list of strings.  If s is the struct corresponding to a row, then I could write something like: 
```
list_transform( struct_keys(s),  k -> if( typeof(s[k]) = 'VARCHAR', s[k], null)).list_filter(x -> x is not null)
```

**Reply by soerenwolfers** [reply]
I'm all for struct_keys!

**Comment by pkoppstein** [comment]
@soerenwolfers - That's great news!

In case you're wondering about why I proposed struct_keys() rather than struct_transform() it's because I could see the latter would likely raise many questions related to types.  The keys of a named struct by contrast are all strings, and the keys of unnamed structs are all integers -- simplicity itself!
